api-put-object-streaming.go

```diff
@@ -0,0 +1,547 @@
+/*
+ * Minio Go Library for Amazon S3 Compatible Cloud Storage (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package minio
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/minio/minio-go/pkg/s3utils"
+)
+
+// PutObjectStreaming using AWS streaming signature V4
+func (c Client) PutObjectStreaming(bucketName, objectName string, reader io.Reader) (n int64, err error) {
+	return c.PutObjectWithProgress(bucketName, objectName, reader, nil, nil)
+}
+
+// putObjectMultipartStream - upload a large object using
+// multipart upload and streaming signature for signing payload.
+// Comprehensive put object operation involving multipart uploads.
+//
+// Following code handles these types of readers.
+//
+//  - *os.File
+//  - *minio.Object
+//  - Any reader which has a method 'ReadAt()'
+//
+func (c Client) putObjectMultipartStream(bucketName, objectName string,
+	reader io.Reader, size int64, metadata map[string][]string, progress io.Reader) (int64, error) {
+
+	if size > 0 && size > minPartSize {
+		// Verify if reader is *os.File, then use file system functionalities.
+		if isFile(reader) {
+			return c.putObjectMultipartStreamFromFile(bucketName, objectName, reader.(*os.File), size, metadata, progress)
+		}
+		// Verify if reader is *minio.Object or io.ReaderAt.
+		// NOTE: Verification of object is kept for a specific purpose
+		// while it is going to be duck typed similar to io.ReaderAt.
+		// It is to indicate that *minio.Object implements io.ReaderAt.
+		// and such a functionality is used in the subsequent code
+		// path.
+		if isObject(reader) || isReadAt(reader) {
+			return c.putObjectMultipartStreamFromReadAt(bucketName, objectName, reader.(io.ReaderAt), size, metadata, progress)
+		}
+	}
+	return c.putObjectMultipartStreamNoChecksum(bucketName, objectName, reader, size, metadata, progress)
+}
+
+// uploadedPartRes - the response received from a part upload.
+type uploadedPartRes struct {
+	Error   error // Any error encountered while uploading the part.
+	PartNum int   // Number of the part uploaded.
+	Size    int64 // Size of the part uploaded.
+	Part    *ObjectPart
+}
+
+type uploadPartReq struct {
+	PartNum int         // Number of the part uploaded.
+	Part    *ObjectPart // Size of the part uploaded.
+}
+
+func (c Client) putObjectMultipartStreamFromFile(bucketName, objectName string,
```

Comment:
> ReadAt is for GetObject, File is for *os.File .

I know, but the code is the same, below is the diff between putObjectMultipartStreamFromFile and putObjectMultipartStreamFromFile (I removed comments before diff)

```diff
@@ -1,4 +1,4 @@
-func (c Client) putObjectMultipartStreamFromReadAt(bucketName, objectName string,
+func (c Client) putObjectMultipartStreamFromFile(bucketName, objectName string,
 	reader io.ReaderAt, size int64, metadata map[string][]string, progress io.Reader) (int64, error) {
 	if err := s3utils.CheckValidBucketName(bucketName); err != nil {
 		return 0, err
@@ -21,10 +21,10 @@
 		return 0, err
 	}
 
-	uploadPartsCh := make(chan uploadPartReq, 10000)
-
 	uploadedPartsCh := make(chan uploadedPartRes, 10000)
 
+	uploadPartsCh := make(chan uploadPartReq, 10000)
+
 	lastPartNumber := totalPartsCount
 
 	for p := 1; p <= totalPartsCount; p++ {
@@ -35,7 +35,6 @@
 	for w := 1; w <= totalWorkers; w++ {
 		go func() {
 			for uploadReq := range uploadPartsCh {
-
 				readOffset := int64(uploadReq.PartNum-1) * partSize
 
 				if uploadReq.PartNum == lastPartNumber {
@@ -46,11 +45,10 @@
 				sectionReader := io.NewSectionReader(reader, readOffset, partSize)
 
 				var objPart ObjectPart
-				objPart, err = c.uploadPart(bucketName, objectName, uploadID, sectionReader,
-					uploadReq.PartNum, nil, nil, partSize)
+				objPart, err = c.uploadPart(bucketName, objectName, uploadID, sectionReader, uploadReq.PartNum,
+					nil, nil, partSize)
 				if err != nil {
 					uploadedPartsCh <- uploadedPartRes{
-						Size:  0,
 						Error: err,
 					}
 					return
@@ -75,7 +73,7 @@
 		}
 		part := uploadRes.Part
 		if part == nil {
-			return 0, ErrInvalidArgument(fmt.Sprintf("Missing part number %d", uploadRes.PartNum))
+			return totalUploadedSize, ErrInvalidArgument(fmt.Sprintf("Missing part number %d", uploadRes.PartNum))
 		}
 		totalUploadedSize += uploadRes.Size
 		if progress != nil {
```